File: SocketTest/java/SocketTestServer.java
package test;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import test.InputMessage;
import test.OutputMessage;
import org.littletonrobotics.junction.Logger;

public class SocketTestServer {
  private static InputMessage inputs = InputMessage.getDefaultInstance();
  private static OutputMessage.Builder outputsBuilder = OutputMessage.newBuilder();

  private static DatagramSocket socket;
  private static InetAddress clientAddress;
  private static int clientPort;
  private static boolean lastConnected = false;

  private static double lastMessageTimestamp = -1.0;
  private static int incomingBytes = 0;
  private static int outgoingBytes = 0;
  private static Timer bandwidthTimer = new Timer(); // TODO fix

  private static final Lock inputsLock = new ReentrantLock();
  private static final Condition newInputsCondition = inputsLock.newCondition();

  public static void start() {
    if (socket != null)
      return;

    try {
      socket = new DatagramSocket(Constants.serverPort);
    } catch (SocketException e) {
      e.printStackTrace();
    }

    if (socket != null) {
      var serverThread = new Thread(
          () -> {
            System.out.println(
                "[SocketTestServer] Socket ready on port "
                    + Integer.toString(Constants.serverPort)
                    + ", waiting for client...");
            byte[] buffer = new byte[Constants.maxPacketSizeBytes];
            DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
            while (true) {
              try {
                socket.receive(packet);

                if (clientAddress == null
                    || !clientAddress.equals(packet.getAddress())
                    || clientPort != packet.getPort()) {
                  System.out.println(
                      "[SocketTestServer] Connected to client "
                          + packet.getAddress().getHostAddress()
                          + ":"
                          + packet.getPort());
                  bandwidthTimer.reset();
                }

                clientAddress = packet.getAddress();
                clientPort = packet.getPort();

                var newInputs = InputMessage.parseFrom(
                    ByteBuffer.wrap(packet.getData(), 0, packet.getLength()));

                inputsLock.lock();
                try {
                  inputs = newInputs;
                  lastMessageTimestamp = Timer.getFPGATimestamp();
                  incomingBytes += packet.getLength();
                  newInputsCondition.signalAll();
                } finally {
                  inputsLock.unlock();
                }
              } catch (IOException e) {
                clientAddress = null;
                System.out.println(
                    "[SocketTestServer] Error receiving packet, waiting for reconnection...");
                e.printStackTrace();
              }
            }
          });

      serverThread.setName("SocketTestServer");
      serverThread.setDaemon(true);
      serverThread.start();
      bandwidthTimer.start();
    }
  }

  public static boolean waitForInputMessage(long timeoutUs) {
    long timeoutNanos = (long) (timeoutUs * 1000L);
    inputsLock.lock();
    try {
      double initialTimestamp = lastMessageTimestamp;
      while (lastMessageTimestamp == initialTimestamp) {
        if (timeoutNanos <= 0L) {
          return false;
        }
        timeoutNanos = newInputsCondition.awaitNanos(timeoutNanos);
      }
      return true; // New message was received
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      return false;
    } finally {
      inputsLock.unlock();
    }
  }

  public static boolean isConnected() {
    return clientAddress != null
        && Timer.getFPGATimestamp() - lastMessageTimestamp < (double) Constants.timeoutMs / 1000.0;
  }

  public static InputMessage getInputMessage() {
    inputsLock.lock();
    try {
      return inputs;
    } finally {
      inputsLock.unlock();
    }
  }

  public static void setOutputMessage(String message) {
    outputsBuilder.setMessage(message);
  }

  public static void processIncoming() {
    inputsLock.lock();
    try {
      outputsBuilder.setSeqnum(inputs.getSeqnum());
      if (bandwidthTimer.advanceIfElapsed(1)) {
        System.out.println("InputsMbps: " + (incomingBytes * 8.0 / 1.0e6));
        System.out.println("OutputsMbps: " + (outgoingBytes * 8.0 / 1.0e6));
        incomingBytes = 0;
        outgoingBytes = 0;
      }
    } finally {
      inputsLock.unlock();
    }
  }

  public static void processOutgoing() {
    if (!isConnected()) {
      if (lastConnected) {
        System.out.println("[SocketTestServer] Client timed out, waiting for reconnection...");
      }
      lastConnected = false;
      return;
    }
    lastConnected = true;

    try {
      byte[] payload = outputsBuilder.build().toByteArray();
      DatagramPacket packet = new DatagramPacket(payload, payload.length, clientAddress, clientPort);
      socket.send(packet);
      outgoingBytes += payload.length;
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  private SocketTestServer() {
  }
}

File: SocketTest/cpp/SocketTestServer.cpp
#include "SocketTestServer.h"

#include <unistd.h>

#include <chrono>
#include <iostream>
#include <vector>

#include "Constants.h"

bool SocketTestServer::started = false;
int SocketTestServer::serverSocketFd = -1;
struct sockaddr_in SocketTestServer::clientAddress;
bool SocketTestServer::clientAddressValid = false;
bool SocketTestServer::lastConnected = false;
std::atomic<double> SocketTestServer::lastMessageTimestamp{-1.0};
std::mutex SocketTestServer::inputsMutex;
std::condition_variable SocketTestServer::newInputsCondition;
test::InputMessage SocketTestServer::inputs = test::InputMessage();
test::OutputMessage SocketTestServer::outputs = test::OutputMessage();

void SocketTestServer::runServer() {
  std::cout << "[SocketTestServer] Socket ready on port " << Constants::serverPort
            << ", waiting for client..." << std::endl;
  std::vector<uint8_t> buffer(Constants::maxPacketSizeBytes);
  struct sockaddr_in currentClientAddr;
  socklen_t clientAddrLen = sizeof(currentClientAddr);

  while (true) {
    ssize_t bytesReceived =
        recvfrom(serverSocketFd, buffer.data(), buffer.size(), 0,
                 (struct sockaddr *)&currentClientAddr, &clientAddrLen);

    if (bytesReceived > 0) {
      if (!clientAddressValid ||
          clientAddress.sin_addr.s_addr != currentClientAddr.sin_addr.s_addr ||
          clientAddress.sin_port != currentClientAddr.sin_port) {
        char clientIpStr[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &(currentClientAddr.sin_addr), clientIpStr,
                  INET_ADDRSTRLEN);
        std::cout << "[SocketTestServer] Connected to client " << clientIpStr
                  << ":" << ntohs(currentClientAddr.sin_port) << std::endl;
      }
      clientAddress = currentClientAddr;
      clientAddressValid = true;

      test::InputMessage newInputs;
      if (newInputs.ParseFromArray(buffer.data(), bytesReceived)) {
        std::unique_lock<std::mutex> lock(inputsMutex);
        inputs = newInputs;
        lastMessageTimestamp.store(
            std::chrono::duration<double>(
                std::chrono::system_clock::now().time_since_epoch())
                .count());
        lock.unlock();
        newInputsCondition.notify_all();
      } else {
        std::cerr << "[SocketTestServer] Failed to parse input message"
                  << std::endl;
      }
    } else if (bytesReceived < 0) {
      perror("[SocketTestServer] Error receiving packet");
      clientAddressValid = false;
    }
  }
}

void SocketTestServer::start() {
  if (started)
    return;
  started = true;

  serverSocketFd = socket(AF_INET, SOCK_DGRAM, 0);
  if (serverSocketFd < 0) {
    perror("[SocketTestServer] FATAL: Socket creation failed");
    return;
  }

  struct sockaddr_in serverAddr;
  serverAddr.sin_family = AF_INET;
  serverAddr.sin_addr.s_addr = INADDR_ANY;
  serverAddr.sin_port = htons(Constants::serverPort);

  if (bind(serverSocketFd, (struct sockaddr *)&serverAddr,
           sizeof(serverAddr)) < 0) {
    perror("[SocketTestServer] FATAL: Bind failed");
    close(serverSocketFd);
    serverSocketFd = -1;
    return;
  }

  std::thread serverThread(runServer);
  if (serverThread.joinable()) {
    serverThread.detach();
  }
}

bool SocketTestServer::waitForInputMessage(long timeoutUs) {
  std::unique_lock<std::mutex> lock(inputsMutex);
  double initialTimestamp = lastMessageTimestamp.load();

  auto status = newInputsCondition.wait_for(
      lock, std::chrono::microseconds(timeoutUs),
      [&] { return lastMessageTimestamp.load() != initialTimestamp; });

  return status; // Returns true if new message was received
}

bool SocketTestServer::isConnected() {
  if (!clientAddressValid || lastMessageTimestamp.load() < 0) {
    return false;
  }
  double now = std::chrono::duration<double>(
                   std::chrono::system_clock::now().time_since_epoch())
                   .count();
  return (now - lastMessageTimestamp.load()) <
         (static_cast<double>(Constants::timeoutMs) / 1000.0);
}

const test::InputMessage SocketTestServer::getInputMessage() {
  std::lock_guard<std::mutex> lock(inputsMutex);
  return inputs;
}

void SocketTestServer::setOutputMessage(const std::string &message) {
  outputs.set_message(message);
}

void SocketTestServer::processIncoming() {
  std::lock_guard<std::mutex> lock(inputsMutex);
  outputs.set_seqnum(inputs.seqnum());
}

void SocketTestServer::processOutgoing() {
  if (!isConnected()) {
    if (lastConnected) {
      std::cout
          << "[SocketTestServer] Client timed out, waiting for reconnection..."
          << std::endl;
    }
    lastConnected = false;
    return;
  }
  lastConnected = true;

  std::string serializedOutputData;
  if (!outputs.SerializeToString(&serializedOutputData)) {
    std::cerr << "[SocketTestServer] Failed to serialize outputs message"
              << std::endl;
    return;
  }

  ssize_t bytesSent =
      sendto(serverSocketFd, serializedOutputData.data(),
             serializedOutputData.length(), 0,
             (struct sockaddr *)&clientAddress, sizeof(clientAddress));
  if (bytesSent < 0) {
    perror("[SocketTestServer] Error sending packet");
  }
}